### PR TITLE
refactor (actions): make better use of variables, secrets and versions

### LINKED ISSUE #3359: Refactor GitHub Actions for flexibility, readability and integrity
## Motivation

- Actual GitHub Actions do not consider triggering depending on the files being changed, and some PR's might go through unneeded tests.
- Some dynamic values are hardcoded, those should be moved to secrets
- Some jobs should have dependencies, for example...if a build fails, there's no need to also wait for testing.
- Linting and styling should have their own Github Action file (other things might also apply)

### Designs

- Create dependencies where needed
- Use secrets for GCP $PROJECT_ID, $ZONE, and other variables for greater flexibility and fewer commits in case of changes.
- Use rlespinasse/github-slug-action@v3.x for more flexible use of variables from Github (repo name, branch name, SHA, author, etc.)
- Segment each workflow objective depending on the action being executed (Push, PR, Release, other) or the files being changed

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

**ENV and Secrets**: Using regex to extract information from GitHub environments is noisy and error-prone, and it restrict flexibility or increases complexity when more specific versions of specific variables are needed, for example, using a cleaned version of the branch name without special characters and with a specific max length (to be used in URL or image tags). 

**Actions versions**: As actions get recurrent fixes, using a specific version causes more maintenance on the pipelines. And using `master` versions could make some action unreliable, as breaking changes might be included without further notice, and even change behavior on a daily basis.

### Designs
A whole step with regex was being used to extract different variables from GitHub's environment. This gets deprecated in favor of using `rlespinasse/github-slug-action@v4` which has slug URL variables.

A SLUG on a variable will:
- put the variable content in lower case
- replace any character by - except 0-9, a-z, ., and _
- remove leading and trailing - character
- limit the string size to 63 characters

This change also takes care of using the Head or Base branch for deployments. This will allow us to potentially merge workflows, as most steps on this deployment actions are very similar, with little variations between workflows.

Related: #3359 

## Solution

- Remove repeated instances of global environment variables.
- Do not print ENV variables on the terminal as GitHub Actions already shows it.
- Use secrets for sensitive information
- Use fixed major versions for actions
- Implement `rlespinasse/github-slug-action@v4`

## Review
@conradoplg @dconnolly @teor2345 

### Reviewer Checklist
  - [x] Validate design
  - [x] Check for Errors

## Follow Up Work
Accomplish the same behavior of CD, Manual Deploy, Regenerate test state, Test and Zcashd Manual Deploy with fewer YAMLs

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -46,89 +39,64 @@ jobs:
...
           service_account_key: ${{ secrets.GCLOUD_AUTH }}
 
-      # Run once: create firewall rule to allow incoming traffic to the nodes
-      # - name: Create Zcash incoming traffic firewall rule
-      #   run: |
-      #     gcloud compute firewall-rules create "allow-zcash" \
-      #     --target-tags zebrad \
-      #     --allow tcp:8233,tcp:18233 \
-      #     --source-ranges 0.0.0.0/0 \
-      #     --description="Allow incoming Zcash traffic from anywhere" \
```
Comment (teor2345) [inline .github/workflows/cd.yml:85]: Can you split this out into a manually run job or script, instead of deleting it?

**Comment by teor2345** [comment]
I think @dconnolly and @conradoplg have the most experience with this CI configs, so I'll leave the reviews to them.

I just want to make sure we're going to check all these jobs work after the refactor?
(Particularly the ones that are only run manually.)

File: .github/workflows/cd.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml:85]: I'll create an issue with this, as a better place for these "one-timers" is a repo specifically for infrastructure configuration. I'll be creating that repo and working with this and Terraform after finishing with the whole CI/CD pipeline improvement.

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -7,36 +7,29 @@ on:
...
-    - name: Set project and image names
-      run: |
-        BRANCH_NAME=$(expr $GITHUB_REF : '.*/\(.*\)') && \
-        BRANCH_NAME=${BRANCH_NAME,,} && \
-        REPOSITORY=${GITHUB_REPOSITORY,,} && \
-        echo "BRANCH_NAME=$BRANCH_NAME" >> $GITHUB_ENV && \
-        echo "SHORT_SHA=$(git rev-parse --short=7 $GITHUB_SHA)" >> $GITHUB_ENV && \
-        echo "REPOSITORY=$REPOSITORY" >> $GITHUB_ENV
+    - name: Inject slug/short variables
+      uses: rlespinasse/github-slug-action@v4
```
Comment (dconnolly) [inline .github/workflows/cd.yml:27]: Nice

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -46,89 +39,64 @@ jobs:
...
+          gcloud compute instance-groups list | grep "zebrad-${{ env.GITHUB_HEAD_REF_SLUG || env.GITHUB_REF_SLUG }}" | grep "${{ env.REGION }}"
 
       # Deploy new managed instance group using the new instance template
       - name: Create managed instance group
         if: steps.does-group-exist.outcome == 'failure'
         run: |
           gcloud compute instance-groups managed create \
-          "zebrad-$BRANCH_NAME" \
-          --template "zebrad-$BRANCH_NAME-$SHORT_SHA" \
+          "zebrad-${{ env.GITHUB_HEAD_REF_SLUG || env.GITHUB_REF_SLUG }}" \
```
Comment (dconnolly) [inline .github/workflows/cd.yml:88]: Can this be set once and reused by the rest of the workflow?

**Comment by dconnolly** [comment]
> Actions versions: As actions get recurrent fixes, using a specific version causes more maintenance on the pipelines. And using master versions could make some action unreliable, as breaking changes might be included without further notice, and even change behavior on a daily basis.

Agreed on moving away from action@master/main. The other actions are pinned to specific versions but only because we have Dependabot configured to propose Action updates as they become available

File: .github/workflows/cd.yml (hunk 3 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml:88]: That would require a bit more refactoring in this workflow. I'm working on another branch on the refactor, so we'll have a PR with this recommendation soon enough.

**Review by conradoplg — COMMENTED** [review]
Looks good!

As @dconnolly mentioned I think we can keep referring to specific versions and bump them with dependabot, so that we can review the changes in the actions before using new versions. I'll leave approval to @dconnolly

**Comment by gustavovalverde** [comment]
@dconnolly reverted to specific action versions for approval

**Review by dconnolly — APPROVED** [review]
LGTM!

**Comment by dconnolly** [comment]
@gustavovalverde the Google Cloud Build may have timed out again, I've kicked it off, if it fails again maybe rebase this branch onto main the get the timeout bump

**Comment by teor2345** [comment]
> @gustavovalverde the Google Cloud Build may have timed out again, I've kicked it off, if it fails again maybe rebase this branch onto main the get the timeout bump

I've opened #3416 for this bug, and nominated it for the stable release candidate epic.

Can you please copy and paste one or two failure logs into that ticket?
And tag #3416 every time it happens?

**Pull request merged**